Create IndexedDB version directory in network process

szewai opened the pull request:
#### bd9231efecf2d6baba24f3b2d65846a90c9259dc
<pre>
Create IndexedDB version directory in network process
<a href="https://bugs.webkit.org/show_bug.cgi?id=270192">https://bugs.webkit.org/show_bug.cgi?id=270192</a>
<a href="https://rdar.apple.com/123725258">rdar://123725258</a>

Reviewed by Geoffrey Garen and Chris Dumez.

We currently create a v0 directory under IndexedDB root directory on main thread of UI process. This patch moves that to
the storage queue in network process to avoid blocking UI process. Also, instead of creating a symlink that points to
the root directory, we create an actual directory and move files into it, because the version upgrade was shipped over
4 years ago and most data should already be migrated to v1 directory. We don&apos;t have to keep the circular symlink now.
This patch also removes the v0 directory if it is not needed.

This change is covered by existing IndexedDB API tests (e.g IndexedDB.IndexedDBFileNameV0) as they are using different
versions of data.

* Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
(WebKit::IDBStorageManager::createVersionDirectoryIfNeeded):
* Source/WebKit/NetworkProcess/storage/IDBStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::NetworkStorageManager):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::addSession):
(WebKit::NetworkProcessProxy::createSymLinkForFileUpgrade): Deleted.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:

Canonical link: <a href="https://commits.webkit.org/275555@main">https://commits.webkit.org/275555@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1176faa2e50c4c9ad6034fac590e7cb516502d45

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21146 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44522 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44716 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38242 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24335 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18476 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34882 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42702 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36279 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15816 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15742 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46150 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38318 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37642 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41532 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16944 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13929 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40092 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18563 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9442 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18624 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18208 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
```diff
@@ -66,6 +66,56 @@ static bool migrateOriginDataImpl(const String& oldOriginDirectory, const String
+    bool oldVersionDirectoryExists = false;
+    Vector<String> oldVersionNames;
+    auto fileNames = FileSystem::listDirectory(rootDirectory);
+    for (auto name : fileNames) {
```
`auto&` ?

cdumez commented on Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
```diff
@@ -66,6 +66,56 @@ static bool migrateOriginDataImpl(const String& oldOriginDirectory, const String
+    if (!oldVersionDirectoryExists)
+        FileSystem::makeAllDirectories(oldVersionDirectory);
+
+    for (auto name : oldVersionNames) {
```
`auto&` ?

geoffreygaren commented on Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
```diff
@@ -66,6 +66,56 @@ static bool migrateOriginDataImpl(const String& oldOriginDirectory, const String
+    // Delete file or symlink named v0.
+    auto oldVersionDirectory = FileSystem::pathByAppendingComponent(rootDirectory, "v0"_s);
+    if (auto fileType = FileSystem::fileType(oldVersionDirectory)) {
+        if (*fileType == FileSystem::FileType::Regular) {
```
Let's comment that this is probably not possible in normal operation.

geoffreygaren commented on Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
```diff
@@ -66,6 +66,56 @@ static bool migrateOriginDataImpl(const String& oldOriginDirectory, const String
+            FileSystem::deleteFile(oldVersionDirectory);
+            oldVersionDirectoryExists = false;
+        }
+        if (*fileType == FileSystem::FileType::SymbolicLink) {
```
else if?

geoffreygaren reviewed (approved):
r=me

szewai committed:
Create IndexedDB version directory in network process
https://bugs.webkit.org/show_bug.cgi?id=270192
rdar://123725258

Reviewed by Geoffrey Garen and Chris Dumez.

We currently create a v0 directory under IndexedDB root directory on main thread of UI process. This patch moves that to
the storage queue in network process to avoid blocking UI process. Also, instead of creating a symlink that points to
the root directory, we create an actual directory and move files into it, because the version upgrade was shipped over
4 years ago and most data should already be migrated to v1 directory. We don't have to keep the circular symlink now.
This patch also removes the v0 directory if it is not needed.

This change is covered by existing IndexedDB API tests (e.g IndexedDB.IndexedDBFileNameV0) as they are using different
versions of data.

* Source/WebKit/NetworkProcess/storage/IDBStorageManager.cpp:
(WebKit::IDBStorageManager::createVersionDirectoryIfNeeded):
* Source/WebKit/NetworkProcess/storage/IDBStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::NetworkStorageManager):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::addSession):
(WebKit::NetworkProcessProxy::createSymLinkForFileUpgrade): Deleted.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:

Canonical link: https://commits.webkit.org/275555@main
